---
title: "RNA seq batch correction using metadata"
author: "Nick Hirschmuller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
always_allow_html: yes
output:
  html_document:
    word_document:
    toc: yes
    toc_depth: '3'
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: false
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, warning = F, error = F, message = F,
    fig.align = "center", cache = F, dpi = 150
)
```

## **Introduction**
After doing some initial QC we detect strong batch effects try to correct them using information from the meta data.

```{r, load libraries}
library(DESeq2)
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
library(magrittr)
library(PCAtools)
library(umap)
library(patchwork)
library(sva)
library(clusterProfiler)
library(RUVSeq)
library(httpgd)
library(tidyverse)
library(Rtsne)

library(here)
source(here("plot_theme.R"))
source(here("helper_functions.R"))
```

```{r}
calc_pairwise_cor <- function(mat, outliers, condition_){
    keep_samples <- mdata %>% filter(group == condition_)  %>% pull(sampleid)
    tmp_cnts <- mat[, keep_samples]
    tmp_cnts <- tmp_cnts[, !colnames(tmp_cnts) %in% outliers]
    pairwise_correlation <- cor(tmp_cnts, method = "spearman")
    h <- Heatmap(pairwise_correlation,
        show_row_names = F,
        show_column_names = F,
        show_row_dend = F,
        show_column_dend = F,
        col = col_fun,
        name="R"
    ) %>%
        ggplotify::as.ggplot()
    return(list(
        heatmap = h,
        correlation_matrix = pairwise_correlation
    ))
}
```

## Investigate structure of the metadata and experimental setup

```{r}
# load in DESeq2dataset where the libsize outliers are already removed.
# for creation see script **01_initial_QC.rmd**
dds <- readRDS(here("QC", "results", "dds_libsize_outliers_removed.rds"))
mdata <- colData(dds) %>% data.frame()

# investigate the structure of the mdata

# oxldlset corresponds to the different stimulation assays, i.e. they are the biological batches
oxldlset_tally <- mdata %>%
    group_by(line, oxldlset) %>%
    tally() %>%
    arrange()
oxldlset_tally %>% filter(line %in%
    names(table(oxldlset_tally$line)[table(oxldlset_tally$line) > 1]))

# -> there is only a single cell line (donor) [vaiy4] which was split over 2 biological batches.
# (for the other ones that are split we have technical replicates)
# This means by adjusting for cell line, we are effectively also adjusting for biological batch (oxldlset)


# batch1id corresponds to the sequencing batch, i.e. which samples were sequenced together.
# this does not mean that they were differentiated and treated together (this is oxldlset)
batch1id_tally <- mdata %>%
    group_by(line, batch1id) %>%
    tally() %>%
    arrange()

batch1id_tally %>% filter(line %in%
    names(table(batch1id_tally$line)[table(batch1id_tally$line) > 1]))
# 
# there are a few cell lines which were sequenced in different batches:
# mita2, naju1, qolg1, uenn3
# other than that, all cell lines were sequenced in same batch which means that we also capture variance due to batch

pool_tally <- mdata %>%
    group_by(line, pool) %>%
    tally() %>%
    arrange()

pool_tally %>% filter(line %in%
    names(table(pool_tally$line)[table(pool_tally$line) > 1]))

# for pool it is very similar:
# mita2, naju1, qolg1, qunz2, uenn3

# All batch effects due to age, ethnicity, and sex will automatically be adjusted, 
# because we have a paired design
```

## Assess the effect of cell line on expression
```{r}
#######
# PCA #
#######
pca_data <- PCAtools::pca(assay(vst(dds)),
    metadata = colData(dds),
    removeVar = 0.4
)
pca_plot_df <- pca_data$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))
for (i in 1:20) {
    pca_plot_df[, str_interp("PC${i}_explained_var")] <- pca_data$variance[i]
}

# for the PCA plot we only show these, because line has way to many levels, would be very messy.
columns_to_plot <- c("state", "condition", "sex", "ethnicity", "age", "batch1id")
PC_pairings <- lapply(1:3, function(x){
    c(2*x-1,2*x)
})

for (i in PC_pairings) {
    print(str_interp("Working on PC${i[1]} and PC${i[2]}"))
    plots <- lapply(columns_to_plot, function(x) {
        expl_var1 <- round(pca_plot_df[1, str_interp("PC${i[1]}_explained_var")], 2)
        expl_var2 <- round(pca_plot_df[1, str_interp("PC${i[2]}_explained_var")], 2)
        p <- density_scatter(
            df = pca_plot_df,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = x,
            density_plot_ratio = 0.15,
            pt_size = 1.2,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )
        p <- p +
            plot_annotation(
                title = x,
                theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
            )
        return(p)
    })
    final_plot <- cowplot::plot_grid(plotlist = plots, ncol = 3)
    ggsave(
        plot = final_plot,
        filename = here(
            "QC", "results", "imgs", "regressing_line", str_interp("PCA_all_covariates_PC${i[1]}_PC${i[2]}.png")
        ),
        dpi = 150,
        width = 18,
        height = 11.5,
    )
}

################
# BOXPLOTS PCA #
################
columns_to_plot <- c("batch1id", "line", "oxldlset")
boxplots_pca <- lapply(setNames(columns_to_plot, columns_to_plot), function(x) {
    p <- plot_categorical_covar(
        df = pca_plot_df,
        covar = x,
        PC = paste0("PC", 1:4),
        bonferroni_correction = TRUE
    ) %>% wrap_plots(., ncol = 2) &
        theme(
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            axis.text.y = element_text(size=9, face="bold"),
            axis.title.y = element_text(size=10, face="bold")
        )
    p <- p +
        plot_annotation(
            title = x,
            theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
        )
    return(p)
})
# there we can see that batch1id has a big influence. 
boxplots_pca$batch1id

# however, it looks like line has an even bigger influence
boxplots_pca$line

# the same is the case for oxldl
boxplots_pca$oxldlset


##################
# CALCULATE UMAP #
##################
umap_data <- umap(t(norm_data))
umap_plot_df <- umap_data$layout %>%
    data.frame() %>%
    magrittr::set_colnames(c("UMAP1", "UMAP2")) %>%
    rownames_to_column("sampleid") %>% 
    left_join(., mdata)


################
# BOXPLOTS PCA #
################
columns_to_plot <- c("batch1id", "line", "oxldlset")
boxplots_umap <- lapply(setNames(columns_to_plot, columns_to_plot), function(x) {
    p <- plot_categorical_covar(
        df = umap_plot_df,
        covar = x,
        PC = paste0("UMAP", 1:2),
        bonferroni_correction = TRUE
    ) %>% wrap_plots(., ncol = 2) &
        theme(
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            axis.text.y = element_text(size=9, face="bold"),
            axis.title.y = element_text(size=10, face="bold")
        )
    p <- p +
        plot_annotation(
            title = x,
            theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
        )
    return(p)
})

boxplots_umap$batch1id
boxplots_umap$line
boxplots_umap$oxldlset


###########
# HEATMAP #
###########
association_output <- covar_correlation(
    df = pca_plot_df,
    covars = c("pool", "state", "condition", "oxldlset", "sex", "ethnicity", "age", "batch1id", "libsize"),
    components = paste0("PC", 1:15)
)

png(here("QC", "results", "imgs", "all_samples","association_heatmap.png"), res=150, width=9, height=5.5, units="in")
association_output$heatmap
dev.off()

# continue

```


## Adjust count data
```{r}
# regress out donor effect (cell line)
# use limmas removeBatchEffects to emulate what the linear model will eventually do.
mod <- model.matrix(~group, colData(dds))
norm_cnts <- assay(vst(dds))
adjusted_counts <- limma::removeBatchEffect(norm_cnts,
    batch = mdata$line, 
    design = mod
)

#################
# CALCULATE PCA #
#################
pca_data <- PCAtools::pca(adjusted_counts,
    metadata = colData(dds),
    removeVar = 0.4
)
pca_plot_df <- pca_data$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., colData(dds) %>% data.frame(), by = c("sample_id" = "sampleid"))
for (i in 1:20) {
    pca_plot_df[, str_interp("PC${i}_explained_var")] <- pca_data$variance[i]
}

for (i in PC_pairings) {
    print(str_interp("Working on PC${i[1]} and PC${i[2]}"))
    plots <- lapply(c("state", "condition", "sex", "ethnicity", "age", "batch1id"), function(x) {
        expl_var1 <- round(pca_plot_df[1, str_interp("PC${i[1]}_explained_var")], 2)
        expl_var2 <- round(pca_plot_df[1, str_interp("PC${i[2]}_explained_var")], 2)
        p <- density_scatter(
            df = pca_plot_df,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = x,
            density_plot_ratio = 0.15,
            pt_size = 1.2,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )
        p <- p +
            plot_annotation(
                title = x,
                theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
            )
        return(p)
    })
    final_plot <- cowplot::plot_grid(plotlist = plots, ncol = 3)
    ggsave(
        plot = final_plot,
        filename = here(
            "QC", "results", "imgs", "regressing_line", str_interp("PCA_all_covariates_PC${i[1]}_PC${i[2]}_corrected.png")
        ),
        dpi = 150,
        width = 18,
        height = 11.5,
    )
}

##################
# CALCULATE UMAP #
##################
columns_to_plot <- c("state", "condition", "sex", "ethnicity", "age", "batch1id")
umap_data <- umap(t(adjusted_counts))
umap_plot_df <- umap_data$layout %>%
    data.frame() %>%
    magrittr::set_colnames(c("UMAP1", "UMAP2")) %>%
    rownames_to_column("sampleid") %>%
    left_join(., mdata)

umap_plots <- lapply(columns_to_plot, function(x) {
    p <- density_scatter(
        df = umap_plot_df,
        x_variable = "UMAP1",
        y_variable = "UMAP2",
        color_by = x,
        density_plot_ratio = 0.15,
        pt_size = 1.2,
    )
    p <- p +
        plot_annotation(
            title = x,
            theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
        )
    return(p)
})
final_plot <- cowplot::plot_grid(plotlist = umap_plots, ncol = 3)
ggsave(
    plot = final_plot,
    filename = here("QC", "results", "imgs", "regressing_line", "UMAP_all_covariates_corrected.png"),
    dpi = 150,
    width = 18,
    height = 11.5,
)

##################
# CALCULATE TSNE #
##################
columns_to_plot <- c("state", "condition", "sex", "ethnicity", "age", "batch1id")
tsne_data <- Rtsne::Rtsne(t(adjusted_counts))
tsne_plot_df <- tsne_data$Y %>%
    data.frame() %>%
    magrittr::set_colnames(c("TSNE1", "TSNE2")) %>%
    mutate("sampleid" = colnames(adjusted_counts)) %>%
    left_join(., mdata)

tsne_plots <- lapply(columns_to_plot, function(x) {
    p <- density_scatter(
        df = tsne_plot_df,
        x_variable = "TSNE1",
        y_variable = "TSNE2",
        color_by = x,
        density_plot_ratio = 0.15,
        pt_size = 1.2,
    )
    p <- p +
        plot_annotation(
            title = x,
            theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
        )
    return(p)
})

final_plot <- cowplot::plot_grid(plotlist = tsne_plots, ncol = 3)
ggsave(
    plot = final_plot,
    filename = here("QC", "results", "imgs", "regressing_line", "TSNE_all_covariates_corrected.png"),
    dpi = 150,
    width = 18,
    height = 11.5,
)

############
# BOXPLOTS #
############
columns_to_plot <- c("batch1id", "line", "oxldlset")
boxplots_pca <- lapply(setNames(columns_to_plot, columns_to_plot), function(x) {
    p <- plot_categorical_covar(
        df = pca_plot_df,
        covar = x,
        PC = paste0("PC", 1:4),
        bonferroni_correction = TRUE
    ) %>% wrap_plots(., ncol = 2) &
        theme(
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            axis.text.y = element_text(size=9, face="bold"),
            axis.title.y = element_text(size=10, face="bold")
        )
    p <- p +
        plot_annotation(
            title = x,
            theme = theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 20))
        )
    return(p)
})

# looks MUCH better now
boxplots_pca$batch1id
boxplots_pca$line
boxplots_pca$oxldlset

###########
# HEATMAP #
###########
covar_list = c(
    "pool", "state", "condition", "oxldlset", "sex",
    "ethnicity", "age", "batch1id", "libsize", "line"
)
correlation_SVs <- covar_correlation(
    df = pca_plot_df,
    covars = covar_list,
    components = paste0("PC", 1:10)
)
png(here("QC", "results", "imgs", "regressing_line", "association_heatmap_corrected.png"), 
res = 150, width = 9, height = 5.5, units = "in")
correlation_SVs$heatmap
dev.off()
```


### Do ORA to see what PC1 and PC2 correspond to
```{r}
# this is needed for the enrichment analysis that we are doing.
hs_msigdb_df <- msigdbr::msigdbr(species = "Homo sapiens")

gene_sets <- hs_msigdb_df %>%
    filter(gs_subcat == "GO:BP" | gs_cat == "H" | gs_subcat == "CP:KEGG") %>%
    select(gs_name, gene_symbol)


# Extract the genes which most influence PC1 and PC2 and see if they are enriched in any biological domain:
ORA_results <- parallel::mclapply(1:4, function(z) {
    print(paste("Running ORA for PC", z))
    # extract top 300 genes which most influence PC
    top_loadings_gene_names <- pca_data$xvars[order(
        abs(pca_data$loadings[, paste0("PC", z)]),
        decreasing = T
    )[1:300]]

    enrichment_result <- enricher(
        gene = top_loadings_gene_names,
        universe = rownames(dds),
        minGSSize = 5,
        maxGSSize = 300,
        TERM2GENE = gene_sets
    )

    # for each of the biological domains, grep the top 3 sig. terms
    sig_results <- enrichment_result@result %>%
        mutate(domain = str_extract(Description, "^[^_]*")) %>%
        mutate(ratio_numeric = sapply(.$GeneRatio, function(x) {
            eval(
                str2expression(x)
            )
        })) %>%
        filter(p.adjust < 0.05) %>%
        group_by(domain) %>%
        slice_min(., order_by = p.adjust, n = 3) %>%
        mutate(Description_short = shorten_strings(Description, 30))

    if (nrow(sig_results) > 0) {
        enrichment_plot <- ggplot(sig_results, aes(x = ratio_numeric, y = Description_short, color = p.adjust)) +
            geom_segment(aes(
                x = 0, xend = ratio_numeric, y = Description_short, yend = Description_short
            ), color = "black", linewidth = 1.5) +
            geom_point(size = 5) +
            scale_color_gradient(low = "red", high = "blue") +
            theme_Publication_side() +
            ylab("Description") +
            xlab("Gene Ratio") +
            scale_x_continuous(expand = c(0, 0.01))
    } else {
        enrichment_plot <- ggplot() +
            theme_void()
    }
    return(enrichment_plot)
    # ggsave(
    #     plot = enrichment_plot,
    #     file = here("QC", "results", "imgs", str_interp("SVA_ENCODE_${i}"), str_interp("ORA_PC${z}.png")),
    #     dpi=150,
    #     width=10,
    #     height = 6

    # )
}, mc.cores=10)
names(ORA_results) <- paste0("PC",1:4)

# maybe not ideal... PC3 and PC4 are decent though (but not a lot of variance)
ORA_results$PC1
ORA_results$PC2
ORA_results$PC3
ORA_results$PC4
```

### Conclusion
I think that this is better than the SVA approach. Confusing what PC1 and PC2 corresponds to, but other than that looks really decent.
Also, this approach takes the uncertainty out of the number of surrogate variables we need. 





### ENCODE QC recommendations
https://www.encodeproject.org/data-standards/rna-seq/long-rnas/
"Replicate concordance: the gene level quantification should have a Spearman correlation of >0.9 between
isogenic replicates and >0.8 between anisogenic replicates (i.e. replicates from different donors)."

We do not reach this, but I think generally the correlations look quite good! 
We compare group wise PCAs to see if there are any big outliers (there are none).
```{r}
dds <- readRDS(here("QC", "results", "dds_libsize_outliers_removed.rds"))
norm_data <- assay(vst(dds, blind = TRUE))
mdata <- colData(dds) %>% data.frame()
# adjust the counts using limma
mod <- model.matrix(~group, colData(dds))
adjusted_counts <- limma::removeBatchEffect(norm_data,
    batch = dds$line,
    design = mod
)

# we do have some isogenic replicates:
technical_replicate_samples <- mdata %>%
    mutate(samplename = paste(line, group, sep = "_")) %>%
    group_by(samplename) %>%
    mutate(n = n()) %>%
    filter(n == 2) %>%
    arrange(samplename) %>%
    pull(sampleid)

pairings <- lapply(1:(length(technical_replicate_samples)/2), function(x) {
    c(2 * x - 1, 2 * x)
})
pairwise_cor <- sapply(pairings, function(i){
    tmp_samples <- technical_replicate_samples[i]
    cor(adjusted_counts[, tmp_samples[1]], adjusted_counts[, tmp_samples[2]], method="spearman")
})
table(pairwise_cor > 0.9)

# anisogenic replicates
mdata <- mdata %>%
    mutate(group = paste(state, condition, sep="_")) 

cor_result <- lapply(unique(mdata$group), function(x){
    samples <- mdata %>%
        filter(group == x)  %>% 
        pull(sampleid)

    tmp_cnts <- adjusted_counts[, samples]
    cor_map <- cor(tmp_cnts, method = "spearman")
    return(cor_map)
})
names(cor_result) <- unique(mdata$group)

# Now all have a nice correlation!
sapply(cor_result, min)
```


#### M0_ctrl
```{r}
col_fun = colorRamp2(c(0.5, 1), c("blue", "red"))
h <- Heatmap(cor_result$M0_ctrl,
    show_row_names = F,
    show_column_names = F,
    show_row_dend = F,
    show_column_dend = F,
    col = col_fun,
    row_km = 2,
    column_km = 2,
    name="R"
)

h <- draw(h)
# outliers_M0_ctrl <- column_order(draw(h))$`1`
# outliers_M0_ctrl <- rownames(cor_result$M0_ctrl)[outliers_M0_ctrl]
apply(cor_result$M0_ctrl, 2, mean) %>% sort(decreasing = T)

# recalculate PCA with adjusted counts
pca_data_subset <- PCAtools::pca(adjusted_counts[, rownames(cor_result$M0_ctrl)],
    metadata = colData(dds)[rownames(cor_result$M0_ctrl),],
    removeVar = 0.4
)
pca_plot_df_subset <- pca_data_subset$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))

for (i in 1:20) {
    pca_plot_df_subset[, str_interp("PC${i}_explained_var")] <- pca_data_subset$variance[i]
}
i = c(1, 2)
expl_var1 <- round(pca_plot_df_subset[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df_subset[1, str_interp("PC${i[2]}_explained_var")], 2)
density_scatter(
            df = pca_plot_df_subset,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = "batch1id",
            density_plot_ratio = 0.15,
            pt_size = 3,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )

```



#### M1_ctrl
```{r}
col_fun = colorRamp2(c(0.7, 1), c("blue", "red"))
h <- Heatmap(cor_result$M1_ctrl,
    show_row_names = F,
    show_column_names = F,
    show_row_dend = F,
    show_column_dend = F,
    col = col_fun,
    row_km = 2,
    column_km = 2,
    name="R"
)
draw(h)

# looks like there is really just one sample which is very different
outliers_M1_ctrl <- apply(cor_result$M1_ctrl, 2, mean) %>% which.min() %>% names()

# recalculate PCA with adjusted counts
pca_data_subset <- PCAtools::pca(adjusted_counts[, rownames(cor_result$M1_ctrl)],
    metadata = colData(dds)[rownames(cor_result$M1_ctrl),],
    removeVar = 0.4
)
pca_plot_df_subset <- pca_data_subset$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))

for (i in 1:20) {
    pca_plot_df_subset[, str_interp("PC${i}_explained_var")] <- pca_data_subset$variance[i]
}
i = c(1, 2)
expl_var1 <- round(pca_plot_df_subset[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df_subset[1, str_interp("PC${i[2]}_explained_var")], 2)
density_scatter(
    df = pca_plot_df_subset,
    x_variable = str_interp("PC${i[1]}"),
    y_variable = str_interp("PC${i[2]}"),
    color_by = "batch1id",
    density_plot_ratio = 0.15,
    pt_size = 3,
    xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
    ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
)

```

#### M2_ctrl
```{r}
col_fun = colorRamp2(c(0.7, 1), c("blue", "red"))
h <- Heatmap(cor_result$M2_ctrl,
    show_row_names = F,
    show_column_names = F,
    show_row_dend = F,
    show_column_dend = F,
    col = col_fun,
    row_km = 2,
    column_km = 2,
    name="R"
)
draw(h)

# looks like there is really just one sample which is very different
outliers_M2_ctrl <- apply(cor_result$M2_ctrl, 2, mean) %>% which.min() %>% names()



# recalculate PCA with adjusted counts
pca_data_subset <- PCAtools::pca(adjusted_counts[, rownames(cor_result$M2_ctrl)],
    metadata = colData(dds)[rownames(cor_result$M2_ctrl),],
    removeVar = 0.4
)
pca_plot_df_subset <- pca_data_subset$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))

for (i in 1:20) {
    pca_plot_df_subset[, str_interp("PC${i}_explained_var")] <- pca_data_subset$variance[i]
}

i = c(1, 2)
expl_var1 <- round(pca_plot_df_subset[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df_subset[1, str_interp("PC${i[2]}_explained_var")], 2)
density_scatter(
            df = pca_plot_df_subset,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = "batch1id",
            density_plot_ratio = 0.15,
            pt_size = 3,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )
```


#### M0_stim
```{r}
col_fun = colorRamp2(c(0.7, 1), c("blue", "red"))
h <- Heatmap(cor_result$M0_stim,
    show_row_names = F,
    show_column_names = F,
    show_row_dend = F,
    show_column_dend = F,
    col = col_fun,
    row_km = 2,
    column_km = 2,
    name="R"
)
draw(h)

# looks like there is really just one sample which is very different
outliers_M0_stimm <- apply(cor_result$M0_stim, 2, mean) %>% which.min() %>% names()



# recalculate PCA with adjusted counts
pca_data_subset <- PCAtools::pca(adjusted_counts[, rownames(cor_result$M0_stim)],
    metadata = colData(dds)[rownames(cor_result$M0_stim),],
    removeVar = 0.4
)
pca_plot_df_subset <- pca_data_subset$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))

for (i in 1:20) {
    pca_plot_df_subset[, str_interp("PC${i}_explained_var")] <- pca_data_subset$variance[i]
}

i = c(1, 2)
expl_var1 <- round(pca_plot_df_subset[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df_subset[1, str_interp("PC${i[2]}_explained_var")], 2)
density_scatter(
            df = pca_plot_df_subset,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = "batch1id",
            density_plot_ratio = 0.15,
            pt_size = 3,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )
```



#### M1_stim
```{r}
col_fun = colorRamp2(c(0.7, 1), c("blue", "red"))
h <- Heatmap(cor_result$M1_stim,
    show_row_names = F,
    show_column_names = F,
    show_row_dend = F,
    show_column_dend = F,
    col = col_fun,
    row_km = 2,
    column_km = 2,
    name="R"
)
draw(h)

# looks like there is really just one sample which is very different
outliers_M1_stimm <- apply(cor_result$M1_stim, 2, mean) %>% which.min() %>% names()


# recalculate PCA with adjusted counts
pca_data_subset <- PCAtools::pca(adjusted_counts[, rownames(cor_result$M1_stim)],
    metadata = colData(dds)[rownames(cor_result$M1_stim),],
    removeVar = 0.4
)
pca_plot_df_subset <- pca_data_subset$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))

for (i in 1:20) {
    pca_plot_df_subset[, str_interp("PC${i}_explained_var")] <- pca_data_subset$variance[i]
}
i = c(1, 2)
expl_var1 <- round(pca_plot_df_subset[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df_subset[1, str_interp("PC${i[2]}_explained_var")], 2)
density_scatter(
            df = pca_plot_df_subset,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = "batch1id",
            density_plot_ratio = 0.15,
            pt_size = 3,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )
```



#### M2_stim
```{r}
col_fun = colorRamp2(c(0.7, 1), c("blue", "red"))
h <- Heatmap(cor_result$M2_stim,
    show_row_names = F,
    show_column_names = F,
    show_row_dend = F,
    show_column_dend = F,
    col = col_fun,
    row_km = 2,
    column_km = 2,
    name="R"
)
draw(h)

# This looks just fine. No outliers
# recalculate PCA with adjusted counts
pca_data_subset <- PCAtools::pca(adjusted_counts[, rownames(cor_result$M2_stim)],
    metadata = colData(dds)[rownames(cor_result$M2_stim),],
    removeVar = 0.4
)
pca_plot_df_subset <- pca_data_subset$rotated[, 1:20] %>%
    rownames_to_column("sample_id") %>%
    left_join(., mdata, by = c("sample_id" = "sampleid"))

for (i in 1:20) {
    pca_plot_df_subset[, str_interp("PC${i}_explained_var")] <- pca_data_subset$variance[i]
}

i = c(1, 2)
expl_var1 <- round(pca_plot_df_subset[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df_subset[1, str_interp("PC${i[2]}_explained_var")], 2)
density_scatter(
            df = pca_plot_df_subset,
            x_variable = str_interp("PC${i[1]}"),
            y_variable = str_interp("PC${i[2]}"),
            color_by = "batch1id",
            density_plot_ratio = 0.15,
            pt_size = 3,
            xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
            ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
        )

```


```{r}
sessionInfo()
```





