---
title: "Evaluate NGSCheckmate"
author: "Nick Hirschmuller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
always_allow_html: yes
output:
  html_document:
    word_document:
    toc: yes
    toc_depth: '3'
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: false
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, warning = F, error = F, message = F,
    fig.align = "center", cache = F, dpi = 150
)
```

## **Introduction**
Evaluate the NGSCheckmate results that I when running it on the BAM files from the RNAseq. 


```{r, load libraries}
library(data.table)
library(magrittr)
library(tidyverse)
library(httpgd)

library(here)
source(here("plot_theme.R"))
source(here("helper_functions.R"))
```


### Load and reformat the Checkmate results

```{r}
out_dir <- here("NGSCheckMate_matching","RNA_seq_bams_comparison")
checkmate_res <- fread(file.path(out_dir, "output_all.txt"), data.table = FALSE) %>%
    magrittr::set_colnames(c("sample1", "match", "sample2", "correlation", "depth")) %>%
    mutate(sample1 = str_remove_all(sample1, ".vcf")) %>% 
    mutate(sample2 = str_remove_all(sample2, ".vcf"))


ggplot(checkmate_res, aes(x = depth)) +
    geom_density(alpha = 0.5) +
    theme_Publication()


ggplot(checkmate_res %>% filter(depth>10), aes(x = correlation, fill = match)) +
    geom_density(alpha=0.5)+
    theme_Publication()

# the checkmate results are a bit odd, because there is always missing one match.
# for example if sampleA matches sampleB, sampleB should also match sampleA
checkmate_res_rev <- data.frame(
    sample1 = checkmate_res$sample2,
    match = checkmate_res$match,
    sample2 = checkmate_res$sample1,
    correlation = checkmate_res$correlation,
    depth = checkmate_res$depth
)

checkmate_res_wide <- rbind(checkmate_res, checkmate_res_rev) %>%
    filter(depth > 10) %>%
    filter(match == "matched") %>%
    group_by(sample1) %>%
    summarise(
        matched_partners = paste(sample2, collapse = ", "),
        matched_correlation = paste(correlation, collapse = ", "),
        matched_depth = paste(depth, collapse = ", ")
    ) %>%
    ungroup() %>% # to remove the grouping structure
    mutate(n_partners = str_count(matched_partners, ",") + 1)




xyz %>%
    filter() %>%
    filter() %>%
    filter()


```


### Load our mdata and add the predicted pairs
```{r}
mdata <- data.table::fread(here("sex_inference", "results", "mdata_with_sex_prediction.tsv")) %>%
    data.frame() %>%
    left_join(., checkmate_res_wide, by = c("sampleid" = "sample1"))

```


### Analyse output
```{r}
mdata %>% group_by(line) %>% tally()


``` 


```{r}
mdata <- mdata %>%
    left_join(., checkmate_res_wide, by=c("sampleid" = "sample1"))

table(mdata$n_partners)







``` 
