---
title: "Evaluate the QTLtools MBV"
author: "Nick Hirschmuller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
always_allow_html: yes
output:
  html_document:
    word_document:
    toc: yes
    toc_depth: '3'
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: false
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, warning = F, error = F, message = F,
    fig.align = "center", cache = F, dpi = 150
)

```

## **Introduction**
For each bam file, we ran QTLtools MBV. This results in 1175 output files (one for each bam), where the 
genotype concordance between that BAM file and ALL samples in the vcf is listed. We have 695 samples in our vcf.


```{r, load libraries}
library(dbscan)
library(httpgd)
library(tidyverse)
library(here)
library(ggrepel)

source(here("helper_functions.R"))
source(here("plot_theme.R"))
```



```{r}
results <- list.files(here("QTLtools_MBV", "results"),
    pattern = "_small.mbv.out",
    full.names = TRUE
)

## load meta data
mdata <- SummarizedExperiment::colData(readRDS(here("QC", "results", "dds_sva_incl_line_libsize_cleaned.rds"))) %>%
    data.frame()


# go through each file and run dbscan to identify outliers. this is just a first pass and does not have to be perfect.
# ideally, we want 1 big cluster and 1 outlier (might also be an outlier cluster because some samples in the vcf are related)
eval_results <- lapply(results, function(i) {
    print(which(i==results))
    sampleid <- str_extract(i, "MM_oxLDL\\d+")
    if (!sampleid %in% mdata$sampleid) {
        return(NULL)
    }
    mbv_res <- data.table::fread(i) %>%
        data.frame() %>%
        na.omit()
    
    mdata_donor <- mdata[mdata$sampleid == sampleid, "line"]
    

    db_result <- dbscan(mbv_res %>% select(contains("perc_")), eps = 0.08, minPts = 2)
    mbv_res$cluster <- db_result$cluster
    plot <- ggplot(mbv_res, aes(perc_het_consistent, perc_hom_consistent, color = as.character(cluster))) +
        geom_point() +
        chameleon::scale_color_chameleon() +
        theme_Publication() +
        theme(legend.position = "none") +
        xlab("Fraction concordant heterozygous sites") +
        ylab("Fraction concordanthomozygous sites") +
        ggtitle(paste(sampleid, mdata_donor, sep=" - ")) +
        xlim(0, 1)
    
    # if we have just 2 clusters, the smaller one is the match.
    # if we have just one outlier, then this is the match. Outliers are labeled as "0"
    if (n_distinct(mbv_res$cluster) == 2) {
        match = mbv_res %>% filter(cluster %in% c(0, 2))
    } else if (n_distinct(mbv_res$cluster) == 1) {
        print("just one cluster")
        match = data.frame(SampleID = "?")
    } else {
        print("multimatch")
        # cluster 1 is hopefully the big one
        match = mbv_res %>%
            filter(cluster != 1) %>%
            filter(perc_het_consistent > 0.8)
    }
    

    
    donor <- str_remove_all(match$SampleID, "HPSI\\d+i?p?f?-") %>%
        str_remove_all(., "_\\d") %>%
        unique()
    
    mapping = data.frame("donor" = paste(donor, collapse = " | "), "sampleid" = sampleid)
    
    plot <- plot +
        ggrepel::geom_label_repel(
            data = mbv_res %>%
                filter(SampleID %in% match$SampleID) %>%
                mutate(SampleID = str_remove_all(SampleID, "HPSI\\d+i?p?f?-")),
            aes(label = SampleID),
            max.overlaps = 50
        )

    return(list(
        plot = plot,
        mapping = mapping
    ))
})
names(eval_results) <- str_extract(results, "MM_oxLDL\\d+")


# save all plots as one biiiig PDF :) 
pdf(here("QTLtools_MBV", "results", "imgs", "concordance_plots_small.pdf"), width=8, height=6)
for (i in 1:length(eval_results)){
    print(i)
    print(eval_results[[i]][["plot"]])
}
dev.off()

# MM_oxLDL7599552 - MM_oxLDL7599557: tixi4 and boqx2 donors are very similar. according to mdata this is tixi4
# MM_oxLDL7599540 - MM_oxLDL7599545: again boqx2 and tixi4 are very similar. according to mdata this is boqx2. both are females

# MM_oxLDL7623715 -> only one match (which is fine). But many samples in vcf with 0 concordant heterozygous sites...

mapping_df <- lapply(eval_results, function(i) {
    i[["mapping"]]
}) %>% bind_rows()


mapping_df <- mapping_df %>% left_join(., mdata %>%
    mutate(donor_mdata = str_remove_all(line, "\\d+")) %>%
    select(sampleid, donor_mdata, sex, group),
by = "sampleid"
)

# load in the mdata that has the predicted sex info
sex_mdata <- data.table::fread(here("sex_inference", "results", "mdata_with_sex_prediction.tsv"), data.table=F)
mapping_df <- mapping_df %>%
    left_join(., sex_mdata %>% select(sampleid, predicted_sex, predicted_prob)) %>%
    mutate(predicted_sex = factor(predicted_sex))

# remove all samples where we only had one big cluster.
# this means that we did not have any outliers.
# probably means that the sample was just not covered in the vcf which is likely
mapping_df_filtered <- mapping_df %>%
    filter(donor != "?")

# check how big the overlap between our donors is and the samples in the vcf
# load a random result
vcf_samples <- data.table::fread(results[1]) %>%
    data.frame() %>%
    pull(SampleID) %>%
    str_remove_all(., "HPSI\\d+i?p?f?-") %>%
    str_remove_all(., "_\\d") %>%
    unique()
table(unique(mapping_df$donor_mdata) %in% unique(vcf_samples))

# 85 samples are not public. for ~(510) 85*6 samples we expect one big cluster. we obseve 466 which seems very reasonable.


## extract samples that "misbehave".
# for example missmatch between predicted and mdata or multiple predicted

abnormal <- mapping_df_filtered %>%
    filter((str_count(donor, " | ") > 0) | (donor != donor_mdata))
perfect <- mapping_df_filtered %>%
    filter(donor == donor_mdata ) 

# sex perfectly matches for the good ones
perfect$sex == perfect$predicted_sex

# create a pdf showing the perfect ones

# and create pdf showing the shit ones




```