---
title: "Try sex inference"
author: "Nick Hirschmuller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
always_allow_html: yes
output:
  html_document:
    word_document:
    toc: yes
    toc_depth: '3'
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: false
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, warning = F, error = F, message = F,
    fig.align = "center", cache = F, dpi = 150
)
```

```{r, load libraries}
library(biomaRt)
library(glmnet)
library(DESeq2)
library(tidyverse)
library(data.table)
library(caret)
library(here)
library(httpgd)
library(patchwork)
library(sva)

source(here("plot_theme.R"))
source(here("helper_functions.R"))
```


## **Introduction**
We have downloaded the whole ARCH4S dataset. Using a python script (extract_mdata.ipynb) we have extracted
all samples in that dataset that come from macrophage and that have sex information available. 
Note: We have changed from macrophages to whole blood. Reason for this is that some "macrophage" datasets are 
acutally just using whole blood when you look them up...

Here, we will fit the lasso regression to predict sex. We will exclude "XIST" in our prediction because the
distribution looks bad for our data. There have also been reports that show that XIST expression in HipSci cell lines is dodgy:
(see [this](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8097692/) and [this](https://www.cell.com/cell-reports/pdf/S2211-1247(19)30326-2.pdf))

```{r}
# load in mdata
mdata <- fread(here("sex_inference", "results", "mdata_whole_blood.tsv"))
mdata$sex <- ifelse(mdata$sex == "F", "FEMALE", "MALE")
mdata <- mdata %>% na.omit()

# only keep series ids that have used both males and females
keep_ids <- mdata %>%
    group_by(series_id, sex) %>%
    tally() %>%
    ungroup() %>%
    group_by(series_id) %>%
    tally() %>%
    filter(n == 2) %>%
    pull(series_id)

mdata <- mdata %>%
    filter(series_id %in% keep_ids)

# load in count data
raw_data <- fread(here("sex_inference", "results", "counts_whole_blood.tsv.gz")) 

# only keep samples that pass mdata filter
raw_data <- data.frame(raw_data)[, c("gene", mdata$sampleID)]

# read in our data so we know which genes are "available"
our_dds <- readRDS(here("QC", "results", "dds_final.rds"))

# figure out which genes are shared between the datasets
shared_genes <- intersect(rownames(our_dds), raw_data$gene)

raw_data <- raw_data %>%
    filter(gene %in% shared_genes)

# some genes occure multiple times... no idea why.
# we will just sum them together if this happens
raw_data <- raw_data %>%
    group_by(gene) %>%
    summarise(across(everything(), sum)) %>%
    column_to_rownames("gene") 

# reorder the gene names
raw_data <- raw_data[match(shared_genes, rownames(raw_data)), ]

# create a deseq2 object for easy handling manipulation of the data
dds <- DESeqDataSetFromMatrix(
    countData = raw_data,
    colData = mdata %>% column_to_rownames("sampleID"),
    design = ~1
)
dds <- estimateSizeFactors(dds)

# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3885686/ following publication says that vst actually performs worse 
# compared to normal log for lasso regression. 
# we also observed that sometimes vst seemed to transform a bimodel distribution into a trimodal one
norm_data <- log1p(counts(dds, normalize = T))
```


### Some visualizations
```{r, eval=FALSE}
#################
# CALCULATE PCA #
#################
pca_data <- PCAtools::pca(norm_data,
    metadata = colData(dds),
    removeVar = 0.4
)

pca_plot_df <- pca_data$rotated[, 1:20] %>%
    rownames_to_column("sampleID") %>%
    left_join(., mdata, by = c("sampleID" = "sampleID"))
for (i in 1:20) {
    pca_plot_df[, str_interp("PC${i}_explained_var")] <- pca_data$variance[i]
}


##################
# CALCULATE UMAP #
##################
umap_data <- umap::umap(t(norm_data))
umap_plot_df <- umap_data$layout %>%
    data.frame() %>%
    magrittr::set_colnames(c("UMAP1", "UMAP2")) %>%
    rownames_to_column("sampleID") %>% 
    left_join(., mdata)


i = c(1, 2)
expl_var1 <- round(pca_plot_df[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df[1, str_interp("PC${i[2]}_explained_var")], 2)
pca_plot <- density_scatter(
    df = pca_plot_df,
    x_variable = "PC1",
    y_variable = "PC2",
    color_by = "series_id",
    density_plot_ratio = 0.15,
    pt_size = 1.2,
    xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
    ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
)

ggsave(
    plot=pca_plot,
    filename = here("sex_inference", "results", "PCA_unadjusted.png"),
    dpi=400,
    height=10,
    width=12
)


umap_plot <- density_scatter(
    df = umap_plot_df,
    x_variable = "UMAP1",
    y_variable = "UMAP2",
    color_by = "series_id",
    density_plot_ratio = 0.15,
    pt_size = 1.2,
)


```


### Adjust counts
```{r}
# to keep things as similar as possible we will use sva to adjust batch effect.
# probably better to just regress out the effect of the different series, but we want to process both
# the training and test data (i.e. our expression data) as similar as possible
mod <- model.matrix(~sex, colData(dds))
mod0 <- model.matrix(~1, colData(dds))
# svseq <- svaseq(norm_data, mod, mod0, n.sv = 10)
# saveRDS(svseq, here("sex_inference", "results", "sva_10_training_data.rds"))
svseq <- readRDS(here("sex_inference", "results", "sva_10_training_data.rds"))
SVs <- svseq$sv %>% magrittr::set_colnames(paste0("SV", 1:10))
mdata <- cbind(mdata %>% dplyr::select(-contains("SV")), SVs)
formula <- as.formula(paste("~", paste0("SV", 1:10, collapse = " + ")))
cov_mod <- model.matrix(formula, mdata)[, -1] # this removes the intercept term. See t

# adjust counts using limma
adjusted_counts <- limma::removeBatchEffect(
    norm_data,
    covariates = cov_mod,
    design = mod
)

# recalculate PCA with adjusted counts
pca_data_corrected <- PCAtools::pca(
    adjusted_counts,
    metadata = colData(dds),
    removeVar = 0.4
)

pca_plot_df <- pca_data_corrected$rotated[, 1:20] %>%
    rownames_to_column("sampleID") %>%
    left_join(., mdata, by = c("sampleID" = "sampleID"))
for (i in 1:20) {
    pca_plot_df[, str_interp("PC${i}_explained_var")] <- pca_data_corrected$variance[i]
}


##################
# CALCULATE UMAP #
##################
umap_data_corrected <- umap::umap(t(adjusted_counts))
umap_plot_df <- umap_data_corrected$layout %>%
    data.frame() %>%
    magrittr::set_colnames(c("UMAP1", "UMAP2")) %>%
    rownames_to_column("sampleID") %>% 
    left_join(., mdata)

i = c(1, 2)
expl_var1 <- round(pca_plot_df[1, str_interp("PC${i[1]}_explained_var")], 2)
expl_var2 <- round(pca_plot_df[1, str_interp("PC${i[2]}_explained_var")], 2)

pca_plot_corrected <- density_scatter(
    df = pca_plot_df,
    x_variable = "PC1",
    y_variable = "PC2",
    color_by = "series_id",
    density_plot_ratio = 0.15,
    pt_size = 1.2,
    xlab = str_interp("PC${i[1]}: ${expl_var1}% variance"),
    ylab = str_interp("PC${i[2]}: ${expl_var2}% variance")
)

umap_plot_corrected <- density_scatter(
    df = umap_plot_df,
    x_variable = "UMAP1",
    y_variable = "UMAP2",
    color_by = "series_id",
    density_plot_ratio = 0.15,
    pt_size = 1.2,
)

ggsave(
    plot = pca_plot_corrected,
    filename = here("sex_inference", "results", "PCA_corrected_SV10.png"),
    dpi = 400,
    height = 10,
    width = 12
)

```


### Fit the model with scaling
During training/testing no difference. 

```{r, eval=FALSE}
# remove XIST
adjusted_counts_wo_XIST <- adjusted_counts[-which(rownames(adjusted_counts) == "XIST"), ]

# Split the data into training and test set
set.seed(50)
training_samples <- mdata$sex %>% 
  createDataPartition(p = 0.85, list = FALSE)
train_data  <- t(adjusted_counts_wo_XIST)[training_samples, ]
test_data <- t(adjusted_counts_wo_XIST)[-training_samples, ]
train_mdata <- mdata[training_samples, ]
test_mdata <- mdata[-training_samples, ]

# Prepare the matrices for glmnet
# try out scaleing
preProcValues <- preProcess(train_data, method= c("center", "scale"))
trainTransformed <- predict(preProcValues, train_data)
testTransformed <- predict(preProcValues, test_data)

y_train <- train_mdata$sex
y_test <- test_mdata$sex

# Fit logistic regression using cross validation
cv_fit <- cv.glmnet(trainTransformed, y_train,
    family = "binomial",
    type.measure = "class",
    n.folds = 10,
    alpha = 1 # 1 corresponds to lasso regression
)
plot(cv_fit)


# Best lambda
best_lambda <- cv_fit$lambda.min
# print(best_lambda)


# Final model with lambda.min
# this uses all of the train data
# the cv model above just the best fold, i.e. misses 1/10 of the data
lasso.model <- glmnet(trainTransformed, y_train,
    alpha = 1, 
    family = "binomial",
    lambda = best_lambda
)

# saveRDS(
#     lasso.model,
#     here("sex_inference", "results", "lasso_model_sex_with_scaling.rds")
# )

predictive_genes <- coef(lasso.model) %>%
    as.matrix() %>%
    data.frame() %>%
    filter(s0 != 0) %>%
    arrange(s0) %>%
    rownames()


# Make prediction on test data
probabilities <- lasso.model %>% predict(
    newx = testTransformed,
    type = "response"
)

predicted.classes <- ifelse(probabilities > 0.5, "MALE", "FEMALE")

# Model accuracy
mean(predicted.classes == y_test)



```


### Plot the training data
```{r}
plot_data_train <- train_data[, predictive_genes[-which(predictive_genes == "(Intercept)")]] %>%
    data.frame() %>%
    rownames_to_column("sampleid") %>%
    pivot_longer(., cols = !sampleid, values_to = "norm_cnts", names_to = "gene") %>%
    left_join(., mdata %>% select(sampleid = sampleID, sex), by = "sampleid")

p <- ggplot(plot_data_train, aes(x = sex, y = norm_cnts)) +
    ggbeeswarm::geom_quasirandom(width = 0.15, size = 0.1) +
    facet_wrap(~gene, scales = "free_y") +
    theme_Publication() +
    stat_summary(
        fun = "mean",
        geom = "crossbar",
        color = "salmon",
        width = 0.1,
        aes(group = sex),
        position = position_dodge(width = 0.5)
    )
ggsave(
    plot=p,
    filename = here("sex_inference", "results", "training_boxplots.png"),
    width = 7,
    height=5,
    dpi=350,
    scale=1.35
)

### Waterfall plot of the coefficients
data <- coef(lasso.model) %>%
    as.matrix() %>%
    data.frame() %>%
    filter(s0 != 0) %>%
    arrange(s0) %>%
    magrittr::set_colnames("Coefficient") %>% 
    rownames_to_column("gene") %>% 
    filter(gene != "(Intercept)")



# Create waterfall plot with labels
waterfall_plot <- ggplot(data, aes(
    x = reorder(gene, Coefficient), y = Coefficient,
    fill = reorder(gene, Coefficient), label = gene
)) +
    geom_col(show.legend = TRUE) +
    geom_text(aes(y = Coefficient),
        position = position_dodge(width = 0.9),
        vjust = 0.5,
        hjust=ifelse(data$Coefficient >= 0, -0.1, 1.1),
        fontface = "bold",
        angle=90
    ) +
    scale_fill_manual(values = c(
        rep("#BD1616", sum(data$Coefficient < 0)),
        rep("#239B56", sum(data$Coefficient > 0))
    )) +
    labs(
        title = "Relevant Features for Sex Prediction",
        x = NULL, # Removing x-axis label
        y = "Coefficient",
    ) +
    theme_Publication() +
    theme(
        axis.text.x = element_blank(),
        legend.position = "none"
    ) +
    ylim(c(-1, 1))
waterfall_plot

pos_features <- data %>% slice_max(., order_by = Coefficient, n =5) %>% pull(gene)
neg_features <- data %>% slice_min(., order_by = Coefficient, n =5) %>% pull(gene)
most_pred_features <- c(pos_features, neg_features)
ggsave(
    plot = waterfall_plot,
    filename = here("sex_inference", "results", "waterfallplot_featureImportance_wo_XIST.png"),
    width = 9,
    height = 5,
    dpi = 350,
    scale = 1.35
)
```

### Run model on our data
```{r}
dds_our <- readRDS(here("QC", "results", "dds_final.rds"))
# order the genes like our training data
dds_our <- dds_our[rownames(dds_our) %in% shared_genes, ]
dds_our <- dds_our[match(shared_genes, rownames(dds_our)),]

# do the same preprocessing as the training data
norm_data_our <- log1p(counts(dds_our, normalize=T))

mod <- model.matrix(~sex, colData(dds_our))
mod0 <- model.matrix(~ 1, colData(dds_our))
# svseq <- svaseq(norm_data_our, mod, mod0, n.sv = 10)
# saveRDS(svseq, here("sex_inference", "results", "sva_10_our_data.rds"))
svseq <- readRDS(here("sex_inference", "results", "sva_10_our_data.rds"))
SVs <- svseq$sv %>% magrittr::set_colnames(paste0("SV", 1:10))
mdata_our <- cbind(colData(dds_our) %>% data.frame() %>% select(-contains("SV")), SVs)
# Create the formula dynamically
formula <- as.formula(paste("~", paste0("SV", 1:10, collapse = " + ")))
cov_mod <- model.matrix(formula, mdata_our)[, -1] # this removes the intercept term. See this post: https://support.bioconductor.org/p/9144613/

# adjust the counts using limma
adjusted_counts_our_data <- limma::removeBatchEffect(norm_data_our,
    covariates = cov_mod,
    design = mod
)
adjusted_counts_our_data_transformed <- t(predict(preProcValues, t(adjusted_counts_our_data)))


adjusted_counts_our_data_transformed_wo_XIST <- adjusted_counts_our_data_transformed[-which(rownames(adjusted_counts_our_data_transformed) == "XIST"), ]


probabilities_our_data <- lasso.model %>% predict(
    newx = t(adjusted_counts_our_data_transformed_wo_XIST),
    type = "response"
)
our_predicted_class <- ifelse(probabilities_our_data > 0.5, "Male", "Female")
all(mdata_our$sampleid == rownames(our_predicted_class))
mdata_our$predicted_prob <- probabilities_our_data[,1]
mdata_our$predicted_sex <- our_predicted_class[,1]
table(mdata_our$predicted_sex == mdata_our$sex)

confusion_matrix <- confusionMatrix(factor(mdata_our$predicted_sex), factor(mdata_our$sex))
print(confusion_matrix)

data.table::fwrite(
    mdata_our,
    file = here("sex_inference", "results", "mdata_with_sex_prediction.tsv"),
    sep = "\t",
    row.names = FALSE
)



```


### Investigate feature distribution
```{r}
# Function to generate a density plot
plot_density <- function(data_vector, sex = waiver()) {
    # Create the ggplot density plot
    plot <- ggplot(data = data.frame(x = data_vector, sex = sex), aes(x = x, y = after_stat(scaled), fill = sex)) +
        geom_density(alpha = 0.5) + # You can change the color and transparency
        labs(title = "Density Plot", x = "Data", y = "Density") +
        theme_Publication()

    # Print the plot
    plot
}

distribution_plots <- lapply(predictive_genes[-which(predictive_genes == "(Intercept)")], function(gene) {
    train_data <- trainTransformed[,gene]
    our_data <- adjusted_counts_our_data_transformed[gene, ]
    min_value <- floor(min(train_data, our_data))
    max_value <- as.integer(max(train_data, our_data)) + 1
    train_plot <- plot_density(train_data, y_train) +
        theme(axis.title.x = element_blank()) +
        xlim(c(min_value, max_value)) +
        ggtitle(paste0("Training Data - ", gene))
    our_data_plot <- plot_density(our_data, mdata_our$sex) +
        xlab("Normalized Expression") +
        xlim(c(min_value, max_value)) +
        ggtitle(paste0("Our Data - ", gene))
    combined <- (wrap_plots(train_plot, our_data_plot, ncol = 1) &
        theme(legend.position = "none")) %>% ggplotify::as.ggplot()
    return(combined)
})
names(distribution_plots) <- predictive_genes[-which(predictive_genes == "(Intercept)")]

distribution_plots$TXLNGY
for (i in names(distribution_plots)) {
    ggsave(
        file = here("sex_inference", "results", str_interp("${i}_data_distribution_scaled.png")),
        plot = distribution_plots[[i]],
        width = 14,
        height = 9,
        dpi = 300,
    )
}



plot_density <- function(data_vector) {
    # Create the ggplot density plot
    plot <- ggplot(data = data.frame(x = data_vector), aes(x = x, y = after_stat(scaled))) +
        geom_density(alpha = 0.5, fill="grey") + # You can change the color and transparency
        labs(title = "Density Plot", x = "Data", y = "Density") +
        theme_Publication()

    # Print the plot
    plot
}
p1 <- plot_density(scale(adjusted_counts["XIST",])) +
    theme(axis.title.x = element_blank()) +
    xlim(c(-3, 3)) +
    ggtitle(paste0("Training Data - ", "XIST"))


p2<-plot_density(adjusted_counts_our_data_transformed["XIST",]) +
    xlab("Normalized Expression") +
    xlim(c(-3, 3)) +
    ggtitle(paste0("Our Data - ", "XIST"))


combined <- (wrap_plots(p1, p2, ncol = 1) &
    theme(legend.position = "none")) %>% ggplotify::as.ggplot()

```

### Investigate the mismatches
We do not observe any enrichment of mismatches in any condition/group
```{r}
# Function to generate a density plot
plot_density <- function(data_vector, sex = waiver()) {
    # Create the ggplot density plot
    plot <- ggplot(data = data.frame(x = data_vector, sex = sex), aes(x = x, y = after_stat(scaled), fill = sex)) +
        geom_density(alpha = 0.5) + # You can change the color and transparency
        labs(title = "Density Plot", x = "Data", y = "Density") +
        theme_Publication()
    # Print the plot
    plot
}


plot_prob_distr <- function(line) {
    tmp_line <- line
    tmp_df <- mdata_our %>%
        filter(line == tmp_line)

    if (nrow(tmp_df) < 2) {
        return()
    }

    hipsci_sex <- tmp_df %>%
        pull(sex) %>%
        unique()

    predicted_sex_male <- tmp_df %>%
        filter(predicted_sex == "Male") %>%
        pull() %>%
        length()

    predicted_sex_female <- tmp_df %>%
        filter(predicted_sex == "Female") %>%
        pull() %>%
        length()

    p <- plot_density(tmp_df$predicted_prob, "random") +
        theme(legend.position = "none") +
        ggtitle(str_interp("HipSci line: ${tmp_line}")) +
        scale_fill_manual(values = "lightgrey") +
        xlab("Probability of 'Male'") +
        xlim(c(0, 1)) +
        geom_text(
            x = 0.6,
            y = 1,
            label = str_interp("HipSci Sex:\n${hipsci_sex}"),
            inherit.aes = F,
        ) +
        geom_text(
            x = 0.85,
            y = 0.975,
            label = str_interp("Predicted Sex:\n ${predicted_sex_female} times female,\n${predicted_sex_male} times male"), inherit.aes = F,
        )
    return(p)
}


# plot the per cell line distribution of
lines <- mdata_our %>%
    mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
    filter(sex_mismatch == "mismatch") %>%
    pull(line) %>%
    unique()

per_line_prob_distr <- pbmcapply::pbmclapply(1:length(lines), function(i) {
    tmp_line <- lines[[i]]
    p <- plot_prob_distr(tmp_line)
    return(p)
}, mc.cores = 5)
per_line_prob_distr <- per_line_prob_distr[!sapply(per_line_prob_distr, is.null)]

pdf(here("sex_inference", "results", "per_line_prob_distr.pdf"), width=10)
for (i in 1:length(per_line_prob_distr)){
    print(i)
    print(per_line_prob_distr[[i]])
}
dev.off()


# Some "outlier" samples, for example in yair1 or dipe1
# but also some that are completely "wrong".
# very high confidence of female for all 6 but label is actually male.
# for example curn3, uoxz4, airc66, ffdc11, vopm2, peop4, ...
# some where just uncertain in general for example xiby4 or lizq3 -> all close to 0.5

# in our expression plot, label these extreme outliers.
plot_data_ours <- t(adjusted_counts_our_data)[, data %>%
    slice_max(order_by = abs(Coefficient), n = 5) %>%
    pull(gene)] %>%
    data.frame() %>%
    rownames_to_column("sampleid") %>%
    pivot_longer(., cols = !sampleid, values_to = "norm_cnts", names_to = "gene") %>%
    left_join(., mdata_our %>% select(sampleid, sex, predicted_sex, predicted_prob, line),
        by = "sampleid"
    )




# abnormal_lines <- mdata_our %>%
#     mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
#     filter(sex_mismatch == "mismatch") %>%
#     group_by(line) %>%
#     tally() %>%
#     filter(n == 6) %>%
#     slice_head(n = 5) %>%
#     pull(line)

abnormal_lines = "dard1"

prob_distr <- plot_prob_distr(abnormal_lines) + theme(plot.title = element_blank())
expr_distribution <- ggplot(plot_data_ours, aes(x = sex, y = norm_cnts)) +
    ggbeeswarm::geom_quasirandom(width = 0.15, size = 0.1) +
    ggbeeswarm::geom_quasirandom(
        data = plot_data_ours %>% filter(line %in% abnormal_lines),
        aes(x = sex, y = norm_cnts, color = line),
        inherit.aes = FALSE,
        width = 0.15, size = 1
    ) +
    ylab("Expression\n(norm. logcounts)") +
    xlab("Recorded Sex in meta data") +
    facet_wrap(~gene, scales = "free_y", ncol=3) +
    scale_color_manual(values = "#11d4d4") +
    theme_Publication() +
    theme(legend.position = "none")

(wrap_plots(prob_distr, expr_distribution, ncol = 2) + plot_layout(widths = c(1, 1.5)))  %>% 
    ggsave(
        plot = .,
        file = here("sex_inference", "results", str_interp("${abnormal_lines}_prediction_wo_XIST.png")),
        dpi=300,
        height=7,
        width=13.5
    )

mdata_our %>%
    mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
    filter(sex_mismatch == "mismatch") %>%
    group_by(line) %>%
    tally(name="wrong_label") %>%
    data.frame() %>%
    left_join(., mdata_our %>%
        group_by(line) %>%
        tally(name = "total_count")) %>%
    arrange(total_count)
```



#### Create a PDF with all the "problematic" lines
When investigating this, weird things pop up...
For example the following: 
sucd3 -> 6 times correct label, 1 time male. 
The donor right samples before that ("MM_oxLDL9752047" to "MM_oxLDL9752051") are only 5 (M2_stim misses and sucd3 has M2_stim twice).
voqc2 is female, but according to prediction male (just like the one outlier from sucd3). Maybe a weird slip?

---

Same fir uaqe1: 
7 samples, M2_stim is doubled but misses from voqm2 (which comes directly before it in same oxLDL batch).
voqm2 is female so would match the "wrong" prediction.

---
Same for womt1:
7 samples, M2_stim is doubled but misses from quze6 (which comes directly before it in same oxLDL batch).
quze6 is female so would match the "wrong" prediction.

----
xarj1
12 samples, 6 are perfect males, 6 perfect females. different oxldl sets, I think this is flat out the wrong donor 
for one of the sets. 






```{r}
mdata_our %>%
    mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
    filter(sex_mismatch == "mismatch") %>%
    group_by(line) %>%
    tally(name = "wrong_label") %>%
    data.frame() %>%
    left_join(., mdata_our %>%
        group_by(line) %>%
        tally(name = "total_count")) %>%
    left_join(., mdata_our %>% select(line, oxldlset) %>% distinct()) %>% arrange(oxldlset)


mdata_our %>%
    mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
    filter(sex_mismatch == "mismatch") %>%
    group_by(line) %>%
    tally(name = "wrong_label") %>%
    data.frame() %>%
    left_join(., mdata_our %>%
        group_by(line) %>%
        tally(name = "total_count"))  %>% filter(wrong_label == total_count) %>% filter(total_count>3)

problematic_lines <- mdata_our %>%
    mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
    filter(sex_mismatch == "mismatch") %>%
    group_by(line) %>%
    tally(name="wrong_label") %>%
    data.frame() %>%
    left_join(., mdata_our %>%
        group_by(line) %>%
        tally(name = "total_count")) %>%
    arrange(total_count) %>% pull(line)

plots <- lapply(problematic_lines, function(i) {
    print(i)
    prob_distr <- plot_prob_distr(i)
    if(is.null(prob_distr)){
        prob_distr <- ggplot()
    }
    expr_distribution <- ggplot(plot_data_ours, aes(x = sex, y = norm_cnts)) +
        ggbeeswarm::geom_quasirandom(width = 0.15, size = 0.1) +
        ggbeeswarm::geom_quasirandom(
            data = plot_data_ours %>% filter(line %in% i),
            aes(x = sex, y = norm_cnts, color = line),
            inherit.aes = FALSE,
            width = 0.15, size = 1
        ) +
        ylab("Expression\n(norm. logcounts)") +
        xlab("Recorded Sex in meta data") +
        facet_wrap(~gene, scales = "free_y", ncol = 3) +
        scale_color_manual(values = "#11d4d4") +
        theme_Publication() +
        theme(legend.position = "none")
    p <- (wrap_plots(prob_distr, expr_distribution, ncol = 2) + plot_layout(widths = c(1, 1.5))) 
})

pdf(here("sex_inference", "results", str_interp("problematic_lines.pdf")),
    height = 7, width = 13.5
)
for (i in plots) {
    print(i)
}
dev.off()
```


### Highlight all the mismatches
```{r}
mismatches <- mdata_our %>%
    mutate(sex_mismatch = ifelse(predicted_sex != sex, "mismatch", "match")) %>%
    filter(sex_mismatch == "mismatch") %>%
    pull(sampleid) 

p <- ggplot() +
    ggbeeswarm::geom_quasirandom(
        data = plot_data_ours %>%
            filter(!sampleid %in% mismatches),
        aes(x = sex, y = norm_cnts),
        width = 0.15, size = 0.1
    ) +
    ggbeeswarm::geom_quasirandom(
        data = plot_data_ours %>%
            filter(sampleid %in% mismatches) %>%
            filter(sex == "Male" & predicted_sex == "Female"),
        aes(x = sex, y = norm_cnts, color = "#ff6347"),
        inherit.aes = FALSE,
        width = 0.15, size = 0.5, alpha = 0.9
    ) +
    ggbeeswarm::geom_quasirandom(
        data = plot_data_ours %>%
            filter(sampleid %in% mismatches) %>%
            filter(sex == "Female" & predicted_sex == "Male"),
        aes(x = sex, y = norm_cnts, color = "#11d4d4"),
        inherit.aes = FALSE,
        width = 0.15, size = 0.5, alpha = 0.9
    ) +
    ylab("Expression\n(norm. logcounts)") +
    xlab("Recorded Sex in meta data") +
    facet_wrap(~gene, scales = "free_y") +
    theme_Publication() +
    scale_color_manual(labels = c("Predicted Male", "Predicted Female"), values = c("#11d4d4", "#ff6347")) +
    guides(color = guide_legend("Prediction"))


ggsave(
    plot = p,
    file = here("sex_inference", "results", "expression_mismatches_wo_XIST.png"),
    dpi = 300,
    height = 8.5,
    width = 12
)


```

### Use ’t Hoen et al, 2013 method
```{r}
# get a list of all genes from the Y chromosome but exclude 
# pseudoautosomal regions
# https://www.ensembl.org/info/genome/genebuild/human_PARS.html
# chromosome:GRCh38:Y:1 - 10000 is unique to Y but is a string of 10000 Ns
# chromosome:GRCh38:Y:10001 - 2781479 is shared with X: 10001 - 2781479 (PAR1)
# chromosome:GRCh38:Y:2781480 - 56887902 is unique to Y
# chromosome:GRCh38:Y:56887903 - 57217415 is shared with X: 155701383 - 156030895 (PAR2)
# chromosome:GRCh38:Y:57217416 - 57227415 is unique to Y

library("biomaRt")
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
filterlist <- c("Y:1:10000", "Y:2781480:56887902", "Y:57217416:10000000000000000") # till the end
results = getBM(
    attributes = c("hgnc_symbol", "entrezgene_id", "chromosome_name", "start_position", "end_position", "gene_biotype"),
    filters = c("chromosomal_region"),
    values = filterlist,
    mart = ensembl
)

y_chr_genes <- results %>%
    filter(chromosome_name == "Y") %>%
    filter(gene_biotype %in% c("protein_coding", "lncRNA")) %>%
    pull(hgnc_symbol) %>% unique() 

average_y_expression <- norm_data_our %>%
    as.data.frame() %>%
    rownames_to_column("gene") %>%
    filter(gene %in% y_chr_genes) %>%
    column_to_rownames("gene") %>%
    t() %>%
    rowMeans() %>%
    enframe(value = "avg_y_chr", name = "sampleid") 

XIST_expr <- norm_data_our["XIST", ] %>%
    enframe(value = "XIST_expr", name = "sampleid")

p <- left_join(average_y_expression, XIST_expr) %>%
    left_join(., mdata_our) %>%
    ggplot(., aes(x = XIST_expr, y = avg_y_chr, color = sex)) +
    geom_point() +
    theme_Publication() +
    xlab("Expression of XIST") +
    ylab("Mean expression of\ngenes on chrY")
p
ggsave(
    plot = p,
    file = here("sex_inference", "results", "XIST_vs_avg_y_expr.png"),
    dpi = 350,
    height = 4,
    width = 6.2
)






```