---
title: "Try sex inference using genotype data"
author: "Nick Hirschmuller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
always_allow_html: yes
output:
  html_document:
    word_document:
    toc: yes
    toc_depth: '3'
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: false
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, warning = F, error = F, message = F,
    fig.align = "center", cache = F, dpi = 150
)
```



```{r, load libraries}
library(here)
library(tidyverse)

source(here("plot_theme.R"))
source(here("helper_functions.R"))
```


## **Introduction**
Use the publically available HipSci genotype data to predict the sex. We will use plink for this.

We first have to transform all vcf files into something readable by plink. We do this for all available vcf files.
Might help us with LD estimation later
```{r}
# load the hipsci metadata
donors <- data.table::fread("/rds/project/jmmh2/rds-jmmh2-results/private/oxLDL/HipSci/HipSci_donors.tsv",
    data.table = FALSE
)
donors$id <- donors$Name %>% str_remove_all(., "HPSI-") 

# For a lot of the donors, there are multiple lines each with genotype information.
vcf_files <- list.files("~/work_dir/oxLDL/data/oxLDL_data/HipSci/genotype_array_data_imputed",
    include.dirs = TRUE,
    recursive = TRUE,
    pattern = "\\.vcf\\.gz$",
    full.names = TRUE
)

vcf_file_ids <- sub(".+?-([a-z]{4}\\_?\\d?).*", "\\1", vcf_files)

# all the vcf files that we have are covered by donor meta information :)
stopifnot(all(substr(vcf_file_ids, 1, 4) %in% donors$id))


# now the actual code to transform vcf into pink
out_dir = "/rds/project/jmmh2/rds-jmmh2-results/private/oxLDL/HipSci/genotype_array_data_imputed_plink"
dummy_sex ="/rds/project/jmmh2/rds-jmmh2-results/private/oxLDL/HipSci/genotype_array_data_imputed_plink/dummy_sex.txt"
dummy_ <- pbmcapply::pbmclapply(1:length(vcf_files), function(i) {
    tmp_vcf <- vcf_files[i]
    tmp_vcf_id <- sub(".+?-([a-z]{4}\\_?\\d?).*", "\\1", tmp_vcf)
    system(
        str_interp(
            "/rds/user/nh608/hpc-work/software/plink2/plink2
        --vcf ${tmp_vcf}
        --out ${file.path(out_dir, tmp_vcf_id)}
        --split-par b37
        --set-all-var-ids @_#_\\$1_\\$2
        --set-missing-var-ids @_#_\\$1_\\$2
        --new-id-max-allele-len 150 missing
        --update-sex ${dummy_sex}
        --chr 23-24
        --make-bed"
        ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
        ignore.stdout = TRUE
    )
}, mc.cores = 20)

# remove all missing ids
dummy_ <- pbmcapply::pbmclapply(1:length(vcf_files), function(i) {
    tmp_vcf <- vcf_files[i]
    tmp_vcf_id <- sub(".+?-([a-z]{4}\\_?\\d?).*", "\\1", tmp_vcf)
    system(
        str_interp(
            "/rds/user/nh608/hpc-work/software/plink2/plink2
        --vcf ${tmp_vcf}
        --out ${file.path(out_dir, tmp_vcf_id)}
        --split-par b37
        --set-all-var-ids @_#_\\$1_\\$2
        --new-id-max-allele-len 150 missing
        --update-sex ${dummy_sex}
        --chr 23-24
        --make-bed"
        ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
        ignore.stdout = TRUE
    )
}, mc.cores = 20)


# now merge them all together
# create a tmp file with paths to all the individual bim files
tmp_file = tempfile()
tmp_file_ignore = tempfile()

# plink2 --pmerge-list X bfile
plink_paths <- file.path(
    out_dir,
    list.files( 
        out_dir,
        pattern = "\\.bim$",
    ) %>% sub("([a-z]{4}\\_?\\d?).bim$", "\\1", .)
)
data.table::fwrite(data.frame(plink_paths), tmp_file, row.names = FALSE, col.names = FALSE)
data.table::fwrite(data.frame("."), tmp_file_ignore, row.names = FALSE, col.names = FALSE)

# remove all missing ids
dummy_ <- pbmcapply::pbmclapply(1:length(plink_paths), function(i) {
    tmp_vcf <- plink_paths[i]
    tmp_vcf_id <- basename(tmp_vcf)
    system(
        str_interp(
            "/rds/user/nh608/hpc-work/software/micromamba/bin/plink
        --bfile ${tmp_vcf}
        --out ${file.path(out_dir, tmp_vcf_id)}
        --exclude ${tmp_file_ignore}
        --chr 23-24
        --make-bed"
        ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
        ignore.stdout = TRUE
    )
}, mc.cores = 20)

system(
    str_interp(
        "/rds/user/nh608/hpc-work/software/micromamba/bin/plink 
        --merge-list ${tmp_file}
        --out ${file.path(out_dir, 'merged')}
        --make-bed"
    ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
)


# generate pruned subset
system(
    str_interp(
        "/rds/user/nh608/hpc-work/software/micromamba/bin/plink
        --bfile ${file.path(out_dir, 'merged')}
        --indep-pairphase 20000 2000 0.5
        --chr 23-24
        --out ${file.path(out_dir, 'prune')}"
    ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
)

system(
    str_interp(
        "/rds/user/nh608/hpc-work/software/micromamba/bin/plink 
        --bfile ${file.path(out_dir, 'merged')}
        --extract ${file.path(out_dir, 'prune.prune.in')}
        --make-bed
        --out ${file.path(out_dir, 'ld_pruned_xy')}"
    ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
)

# finally run sex check
system(
    str_interp(
        "/rds/user/nh608/hpc-work/software/micromamba/bin/plink
        --bfile ${file.path(out_dir, 'ld_pruned_xy')}
        --check-sex
        --out ${file.path(out_dir, 'sex_check')}"
    ) %>% str_replace_all(., "[\r\n]", ""), # removes new lines so the command works and it stays readable
)
```



### read in the file with the sex prediction
```{r}
sex_prediction <- data.table::fread("/rds/project/jmmh2/rds-jmmh2-results/private/oxLDL/HipSci/genotype_array_data_imputed_plink/sex_check.sexcheck") %>% data.frame()

ggplot(sex_prediction, aes(x = F)) +
    geom_histogram()+
    theme_Publication()

# assign sex according to F statistic

sex_prediction$predicted_sex <- ifelse(sex_prediction$F<0.25, "Female", "Male")
sex_prediction$line <- str_extract(sex_prediction$IID, "-[a-z]{4}") %>% substr(.,2,5)

out <- left_join(
    sex_prediction %>%
        group_by(line) %>%
        tally(name = "line_occurance"),
    sex_prediction %>%
        group_by(line, predicted_sex) %>%
        tally(),
    by = "line"
) %>%
    mutate(mismatch = ifelse(n!=line_occurance, "mismatch", "match"))

# all lines are assign to the correct one.
out$mismatch %>% table()



our_dds <- readRDS(here("QC", "results", "dds_final.rds"))

mdata_our <- colData(our_dds) %>%
    data.frame() %>%
    mutate(tmp_line = substr(line, 1, 4))

final <- left_join(mdata_our, sex_prediction %>% select(line, predicted_sex) %>% distinct(), by = c("tmp_line" = "line"))

# perfect match
x<-final %>% select(line, sex, predicted_sex) %>% distinct()
table(x$sex == x$predicted_sex, useNA="always")
```